src/cntk-train/src/main/scala/DataConversion.scala

```diff
@@ -106,6 +103,74 @@ object DataTransferUtils {
     data.select(uberCol.as('value))
   }
 
+  def toFloat(data: Array[Double]): Array[Float] = data.map(_.toFloat)
+
+  def convertDatasetToCNTKParquetFormat(data: Dataset[_],
+                                        label: String,
+                                        feats: String,
+                                        labelForm: String,
+                                        featureForm: String,
+                                        weightPrecision: String): DataFrame = {
+    // Dense conversion functions
+    val toDenseArray =
+      if (weightPrecision == CNTKLearner.floatPrecision) {
+        udf({
+          vector: Vector => vector match {
+            case dv: DenseVector => toFloat(dv.toArray)
+            case sv: SparseVector => toFloat(sv.toDense.toArray)
+          }
+        })
+      } else {
+        udf({
+          vector: Vector => vector match {
+            case dv: DenseVector => dv.toArray
+            case sv: SparseVector => sv.toDense.toArray
+          }
+        })
+      }
+    // Sparse format conversion functions
+    val getNZs = udf({
+      vector: Vector => vector match {
+        case dv: DenseVector => dv.toSparse.numNonzeros
+        case sv: SparseVector => sv.numNonzeros
+      }
+    })
+    val getIdxes = udf({
+      vector: Vector => vector match {
+        case dv: DenseVector => dv.toSparse.indices
+        case sv: SparseVector => sv.indices
+      }
+    })
+    val getVals =
+      if (weightPrecision == CNTKLearner.floatPrecision) {
+        udf({
+          vector: Vector => vector match {
+            case dv: DenseVector => toFloat(dv.toSparse.values)
+            case sv: SparseVector => toFloat(sv.values)
+          }
+        })
+      } else {
+        udf({
+          vector: Vector => vector match {
+            case dv: DenseVector => dv.toSparse.values
+            case sv: SparseVector => sv.values
+          }
+        })
+      }
+    // Note: order matters, features column needs to come first
+    val cols = data.select(feats, label).columns.flatMap(name => {
+      if ((name == label && labelForm == CNTKLearner.denseForm) ||
+        (name == feats && featureForm == CNTKLearner.denseForm)) {
+        Seq(toDenseArray(data.col(name)).alias(name))
+      } else {
+        Seq(getNZs(data.col(name)).alias(name + "_size"),
+          getIdxes(data.col(name)).alias(name + "_indices"),
+          getVals(data.col(name)).alias(name) + "_values")
```

Comment:
Are we missing a closing ")" to match "Seq("

Reply:
good catch, the string:
+ "_values"
needs to be inside alias() I believe